lib/parser/create.js

```diff
@@ -315,21 +315,35 @@ export function createParser(config) {
         parser.parseValue = 'parseValue' in options ? Boolean(options.parseValue) : true;
         parser.parseCustomProperty = 'parseCustomProperty' in options ? Boolean(options.parseCustomProperty) : false;
 
-        const { context = 'default', onComment } = options;
+        const { context = 'default', onComment, onToken } = options;
 
         if (context in parser.context === false) {
             throw new Error('Unknown context `' + context + '`');
         }
 
-        if (typeof onComment === 'function') {
+        if (typeof onComment === 'function' || typeof onToken === 'function') {
             parser.forEachToken((type, start, end) => {
                 if (type === Comment) {
+                    if (onComment) {
+                        const loc = parser.getLocation(start, end);
+                        const value = cmpStr(source, end - 2, end, '*/')
+                            ? source.slice(start + 2, end - 2)
+                            : source.slice(start + 2, end);
+
+                        onComment(value, loc);
+                    }
+                    return;
+                }
+
+                if (type === WhiteSpace) {
+                    return;
+                }
+
+                if (onToken) {
                     const loc = parser.getLocation(start, end);
```

Comment:
Can you explain more about this? Are you calculating the location from scratch every time `parser.getLocation()` is called?

In my JSON parser ([Momoa](https://github.com/humanwhocodes/momoa)), the lines and columns are tracked inline with tokenization so there's no additional calculation. [Acorn](https://github.com/acornjs/acorn/) similarly tracks line/column position along with source offset when you pass `locations` as an option to `parse()`. Could we maybe do something similar if `onToken` is present?

From a syntax analysis perspective, the full location is always necessary, so I'd prefer to have that produced automatically rather than adding another (potentially expensive) step after the fact.

Reply:
Lines and columns for the source are computed lazily on the first call to [`getLocation()`](https://github.com/csstree/csstree/blob/76fb90ae0422b9f9bc45bbe008778a7db7ddb213/lib/parser/create.js#L252-L262), using [OffsetToLocation](https://github.com/csstree/csstree/blob/master/lib/tokenizer/OffsetToLocation.js) provided `parse()` is called `positions: true`. The location object has the following structure:

```ts
{
    source: string | null,
    start: {
        offset: number,
        line: number,
        column: number
    },
    end: {
        offset: number,
        line: number,
        column: number
    }
}
```

Such structure allocates at least 106 bytes of memory per token, plus the computational overhead of object creation, values read, and math. This cost applies to every token, even if their location data is never used.

In most cases, token locations remain unused. For example, in your case, locations are only needed when an error or warning is encountered. Computing location data for all tokens in advance would result in unnecessary memory consumption and processing time.

Reply:
Right, I totally understand that reasoning. FWIW, ESLint's location structure is the same (minus `source`). My suggestion is that when someone provides `onToken`, they are explicitly saying they want to pay that cost in order to get the token information.

Separately from that, it looks like location [calculation](https://github.com/csstree/csstree/blob/0afee3d88916a2860b48144537027cfb2f48c080/lib/tokenizer/OffsetToLocation.js#L11) always runs on the entire input from start to finish to get all location information. I think that may be part of the performance penalty as you're definitely computing values you might not need.

In Momoa, for example, I have a [`CharCodeReader` class](https://github.com/humanwhocodes/momoa/blob/main/js/src/char-code-reader.js) that is used to read char codes from the input. As it does so, it automatically tracks the line and column such that you can call `locate()` at any point in time and retrieve the offset, line, and column for that character. I'm wondering if something like that would reduce the cost of calculating locations here.